find_email_addresses.py

```diff
@@ -0,0 +1,147 @@
+# find_email_addresses.py
+# 
+# Written by:       Matt Long
+# Last modified:    10/26/15
+#
+# Purpose:          Given a domain, prints out a list of all email addresses 
+#                   that can be found on the given page as well as its child
+#                   pages.
+
+import sys, os
+import re
+from PyQt4.QtGui import *  
+from PyQt4.QtCore import *  
+from PyQt4.QtWebKit import *  
+
+# Matches anything of form w.x@y.z ('.x' is optional)
+EMAIL_REGEX = '(\w+(\.\w*)*@\w+(\.\w*)+)'
+
+# Used to find relative links in html body
+RELATIVE_LINK_FIND_REGEX = 'href="\/[^"]+"'
+RELATIVE_LINK_CLEAN_REGEX = '\/[^"]+'
+
+# TODO: regex for finding absolute links
+
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
+
+# Code for this class was modified by me but originally found on
+# https://webscraping.com/blog/Scraping-multiple-JavaScript-webpages-with-webkit/
+# Module is licenced under LGPL, so I'm allowed to use it (I'm pretty sure)
+#
+# Class that renders a webpage, allowing JS-rendered elements to load and then
+# be scraped by the tool
+class Render(QWebPage):  
+  def __init__(self, urls, cb):
+    self.app = QApplication(sys.argv)  
+    QWebPage.__init__(self)  
+    self.loadFinished.connect(self._loadFinished)  
+    
+    # list of urls to scrape through - added to during execution
+    self.urls = urls  
+
+    # callback so all HTML bodies don't have to be stored in memory
+    self.cb = cb
+
+    # dictionaries of urls and emails to prevent duplication
+    self.visited_sites = {}
+    self.found_emails = {}
+
+    self.crawl()  
+    self.app.exec_()  
+
+
+  # called after init finishes setup    
+  def crawl(self):  
+    if self.urls:  
+      url = self.urls.pop(0)  
+      self.mainFrame().load(QUrl(url))  
+    else:  
+      self.app.quit()  
+
+
+  # fires after page loads in crawl function          
+  def _loadFinished(self, result):  
+    frame = self.mainFrame()  
+    url = str(frame.url().toString())  
```

Comment:
This looks interesting, does the `toString` method not cast the output of `frame.url()` to a string?